### PR TITLE
GUI to edit "Customer Attributes" and "Customer Address Attributes"

### PR DESCRIPTION
**Pull request by fballiano**
Since the beginning of time we always had to rely on commercial modules (or developers) to add/edit/delete "customer attributes" or "customer address attributes". This PR aims to fix this and give the users the power to manage "customer attributes" the same way they do with "product attributes".

This feature was developed mostly using the code of the "product attributes" edit feature, that's why I've copied the license in the file header.

It's a big PR so suggestions/fixes/additions are welcome.

### Manual testing scenarios (*)
1. open the backend
2. open the menu "Customers -> Attributes"
3. Use the two new masks to add, edit or delete customer attributes or customer address attributes

### Screenshots

The new menus
Schermata 2022-06-28 alle 23 50 12

Customer attributes grid
Schermata 2022-06-28 alle 23 51 34

Add/Edit customer attribute
Schermata 2022-06-28 alle 23 52 02

For customer address it's the same thing so I'll not post screenshots of that.

### CODE REVIEW

**Comment by kiatng** [comment]
I have a similar custom extension, including the _Manage Attribute Sets_ for organizing the attributes for different customer groups.

**Comment by ADDISON74** [comment]
For a long time I have not seen pieces of code that bring new features. Unfortunately, editing Customer Attributes and Customer Address Attributes were missing from Magento and I appreciate that you came up sharing the code.

File: app/code/core/Mage/Adminhtml/controllers/Customer/Address/AttributeController.php [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+            $data['frontend_label'] = (array) $data['frontend_label'];
+            foreach ($data['frontend_label'] as & $value) {
+                if ($value) {
+                    $value = $helper->stripTags($value);
+                }
+            }
+
+            if (!empty($data['option']) && !empty($data['option']['value']) && is_array($data['option']['value'])) {
+                $allowableTags = isset($data['is_html_allowed_on_front']) && $data['is_html_allowed_on_front']
+                    ? sprintf('<%s>', implode('><', $this->_getAllowedTags())) : null;
```
Comment (elidrissidev) [inline app/code/core/Mage/Adminhtml/controllers/Customer/Address/AttributeController.php:154]: `_getAllowedTags` is missing in this controller:

```php
/**
 * Get list of allowed text formatted as array
 *
 * @return array
 */
protected function _getAllowedTags()
{
    return explode(',', Mage::getStoreConfig(self::XML_PATH_ALLOWED_TAGS));
}
```

File: app/code/core/Mage/Adminhtml/controllers/Customer/AttributeController.php [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+            $data['frontend_label'] = (array) $data['frontend_label'];
+            foreach ($data['frontend_label'] as & $value) {
+                if ($value) {
+                    $value = $helper->stripTags($value);
+                }
+            }
+
+            if (!empty($data['option']) && !empty($data['option']['value']) && is_array($data['option']['value'])) {
+                $allowableTags = isset($data['is_html_allowed_on_front']) && $data['is_html_allowed_on_front']
+                    ? sprintf('<%s>', implode('><', $this->_getAllowedTags())) : null;
```
Comment (elidrissidev) [inline app/code/core/Mage/Adminhtml/controllers/Customer/AttributeController.php:154]: `_getAllowedTags` is missing in this controller too.

File: app/code/core/Mage/Adminhtml/controllers/Customer/AttributeController.php [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+            ->_title($this->__('Manage Customer Attributes'));
+
+        if($this->getRequest()->getParam('popup')) {
+            $this->loadLayout('popup');
+        } else {
+            $this->loadLayout()
+                ->_setActiveMenu('customer/attributes')
+                ->_addBreadcrumb(Mage::helper('customer')->__('Customer'), Mage::helper('customer')->__('Customer'))
+                ->_addBreadcrumb(
+                    Mage::helper('customer')->__('Manage Customer Attributes'),
```
Comment (elidrissidev) [inline app/code/core/Mage/Adminhtml/controllers/Customer/AttributeController.php:53]: The newly added lines do not have a translation.

File: app/design/adminhtml/default/default/template/customer/address/attribute/js.phtml [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+            case 'media_image':
+                $('front_fieldset').previous().hide();
+                $('front_fieldset').hide();
+
+                setRowVisibility('is_required', false);
+                setRowVisibility('is_unique', false);
+                setRowVisibility('frontend_class', false);
+                break;
+
+        <?php foreach (Mage::helper('catalog')->getAttributeHiddenFields() as $type=>$fields): ?>
```
Comment (elidrissidev) [inline app/design/adminhtml/default/default/template/customer/address/attribute/js.phtml:172]: This method reads a value with key `attribute_type_hidden_fields` from registry, but from what I see it's not set anywhere in customer page.

File: app/design/adminhtml/default/default/template/customer/attribute/js.phtml [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+            case 'media_image':
+                $('front_fieldset').previous().hide();
+                $('front_fieldset').hide();
+
+                setRowVisibility('is_required', false);
+                setRowVisibility('is_unique', false);
+                setRowVisibility('frontend_class', false);
+                break;
+
+        <?php foreach (Mage::helper('catalog')->getAttributeHiddenFields() as $type=>$fields): ?>
```
Comment (elidrissidev) [inline app/design/adminhtml/default/default/template/customer/attribute/js.phtml:172]: Same comment as before.

File: app/code/core/Mage/Adminhtml/controllers/Customer/AttributeController.php [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+                Mage::getSingleton('adminhtml/session')->addError(
+                    Mage::helper('customer')->__('This attribute cannot be deleted.'));
+                $this->_redirect('*/*/');
+                return;
+            }
+
+            try {
+                $model->delete();
+                Mage::getSingleton('adminhtml/session')->addSuccess(
+                    Mage::helper('customer')->__('The product attribute has been deleted.'));
```
Comment (elidrissidev) [inline app/code/core/Mage/Adminhtml/controllers/Customer/AttributeController.php:310]: This phrase should say customer. It's used in multiple places.

**Review by elidrissidev — CHANGES_REQUESTED** [review]

File: app/code/core/Mage/Adminhtml/controllers/Customer/AttributeController.php (hunk 6 above)
Comment (fballiano) [reply app/code/core/Mage/Adminhtml/controllers/Customer/AttributeController.php]: fixed it thanks!!

File: app/code/core/Mage/Adminhtml/controllers/Customer/AttributeController.php (hunk 3 above)
Comment (fballiano) [reply app/code/core/Mage/Adminhtml/controllers/Customer/AttributeController.php:53]: fixed!

File: app/code/core/Mage/Adminhtml/controllers/Customer/AttributeController.php (hunk 2 above)
Comment (fballiano) [reply app/code/core/Mage/Adminhtml/controllers/Customer/AttributeController.php:169]: added

File: app/code/core/Mage/Adminhtml/controllers/Customer/Address/AttributeController.php (hunk 1 above)
Comment (fballiano) [reply app/code/core/Mage/Adminhtml/controllers/Customer/Address/AttributeController.php:169]: ok I've added the method, but it uses a constant that's `const XML_PATH_ALLOWED_TAGS = 'system/catalog/frontend/allowed_html_tags_list';` so what should we do with that? change it to `system/customer/frontend/allowed_html_tags_list`?

File: app/code/core/Mage/Adminhtml/Block/Customer/Address/Attribute/Edit.php [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+class Mage_Adminhtml_Block_Customer_Address_Attribute_Edit extends Mage_Adminhtml_Block_Widget_Form_Container
+{
+    public function __construct()
+    {
+        $this->_objectId = 'attribute_id';
+        $this->_controller = 'customer_address_attribute';
+
+        parent::__construct();
+
+        if($this->getRequest()->getParam('popup')) {
```
Comment (luigifab) [inline app/code/core/Mage/Adminhtml/Block/Customer/Address/Attribute/Edit.php:36]: You have forgotten a space.

File: app/code/core/Mage/Adminhtml/Block/Customer/Address/Attribute/Edit.php [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+                    'class'     => 'save'
+                ),
+                100
+            );
+        }
+
+        $this->_updateButton('save', 'label', Mage::helper('catalog')->__('Save Attribute'));
+        $this->_updateButton('save', 'onclick', 'saveAttribute()');
+
+        if (! Mage::registry('entity_attribute')->getIsUserDefined()) {
```
Comment (luigifab) [inline app/code/core/Mage/Adminhtml/Block/Customer/Address/Attribute/Edit.php:62]: I think you have added a extra space.

File: app/code/core/Mage/Adminhtml/Block/Customer/Address/Attribute/Edit/Form.php [hunk 9]
```diff
@@ -0,0 +1,12 @@
+<?php
+
```
Comment (luigifab) [inline app/code/core/Mage/Adminhtml/Block/Customer/Address/Attribute/Edit/Form.php:2]: You have forgotten the copy comment (you have added it elsewhere).

File: app/code/core/Mage/Adminhtml/Block/Customer/Address/Attribute/Edit/Tab/Main.php [hunk 10]
```diff
@@ -0,0 +1,71 @@
+<?php
+
```
Comment (luigifab) [inline app/code/core/Mage/Adminhtml/Block/Customer/Address/Attribute/Edit/Tab/Main.php:2]: You have forgotten the copy comment (you have added it elsewhere).

**Review by luigifab — CHANGES_REQUESTED** [review]
Thank you so much for this missing feature. I will test it at production level and deploy it!

File: app/code/core/Mage/Adminhtml/controllers/Customer/Address/AttributeController.php (hunk 1 above)
Comment (elidrissidev) [reply app/code/core/Mage/Adminhtml/controllers/Customer/Address/AttributeController.php:169]: I'm not sure if we ever need that for the customer attributes, I don't think any of them allow html. I'll have to look into that.

**Comment by elidrissidev** [comment]
Why was this PR closed?

**Pull request closed without merging**